Add a simple diurnal controller.

uluyol opened the pull request:
The diurnal controller changes the number of replicas of a replication controller based on a list of times and replica counts or on a sample rate and formula.

More specifically, the diurnal controller takes a list of times and replica counts, and the diurnal_math wrapper can generate these from a sample rate and formula. The diurnal controller should itself be run under a replication controller; an example spec file has been provided.

@thockin

quinton-hoole commented:
@k8s-bot OK to test

quinton-hoole commented:
@k8s-bot add to whitelist

uluyol commented:
I signed it!

erictune commented on contrib/diurnal/example-diurnal-controller.yaml:
```diff
@@ -0,0 +1,27 @@
+        name: diurnalcontroller
+    spec:
+      containers:
+        - command: ["diurnal_math", "-ns", "default", "-labels", "name=redis-slave", "-sample", "300", "-countfn", "10+10*sin(t)"]
```
The interface should not involve users typing "a+b*sin(c*t)".  There are way too many ways to make mistakes.  Instead, the diurnal controller should have parameters like:
 - min
 - peak
 - peak hour of day, in a standard time format.

Note that actual load curves are not sinusoidal.  They have steeper ramp up and ramp down, and flatter peaks and narrower troughs.

The other way to do it is piecewise linear with like a couple of (value, time-of-day) points for each day.

uluyol replied:
The actual controller takes in a list of times and replica counts. diurnal_math is a wrapper which instead lets you provide a formula and sample rate at the request of @thockin. Perhaps I should add another example using that.

At the moment the user also provides some maximum rate (the example currently uses the default value, I'll make it explicit) of count change per minute and the controller will transition from one value to another as fast as possible at that rate. It might be easier to just specify some step size instead.

thockin replied:
I thought it was an interesting demonstration to allow arbitrary functions here, but maybe it needs a better name?  It might be that the really useful functionality is just [](time, value) pairs and good (non-linear?) interpolation between them

erictune replied:
I think "hard to shoot yourself in the foot" definitely trumps "interesting" when configuring your web services.  

The actual curves are not simple sine() functions.   Composing an curve that has the desired shape using c library math functions is going to be very error-prone and unreadable (e.g. how to you make the weekends lower or higher just using c library functions?  You can do it, but you have to think about it way to long for it to be safe.)

I'd start with a list of (ISO 8601 time, amplitude) pairs.  An enhancement would be to allow different values for certain days of the week.

erictune commented:
I don't know why you chose to implement this in a combination of go, python and shell.  Can this be done in 1 or at most 2 languages?

uluyol commented:
The python is needed for the math formulas, but the shell can be converted to python.

vmarmol commented:
Thank you very much for the pull request - we sincerely appreciate the work that goes into sending us a change.

We are currently in a feature-freeze and general code slush as we work toward our 1.0 release.  Our initial triage indicates that this is a non-critical change at this time, so we'd like to leave it open and revisit it after 1.0.  We expect the freeze to wrap up in early to mid July.

If you feel that this triage is incorrect, please respond and let us know why.  We're currently accepting changes that meet any of the following criteria:
   1. Changes directly related to v1.0 issues: https://github.com/GoogleCloudPlatform/kubernetes/issues?q=is%3Aopen+is%3Aissue+milestone%3Av1.0
   2. Changes addressing P0 bugs: https://github.com/GoogleCloudPlatform/kubernetes/issues?q=is%3Aopen+is%3Aissue+label%3Apriority%2FP0
   3. Changes to supported platforms to bring them up to par with the primary development platforms
   4. Docs, user experience, and test related fixes

Thanks again for your contribution!  The community is really what has made this project so exciting.

thockin commented on contrib/diurnal/Makefile:
```diff
@@ -0,0 +1,23 @@
+	godep go test .
+
+build: $(BIN)
+	docker build -t uluyol/kube-diurnal:$(TAG) .
```
Make the prefix a variable

uluyol replied:
Done.

thockin commented on contrib/diurnal/Makefile:
```diff
@@ -0,0 +1,23 @@
+BIN = dc
+
+dc: dc.go
+	godep go build dc.go
```
You might want to build this as static, but if it works, maybe not.

uluyol replied:
Done.

thockin commented on contrib/diurnal/dc.go:
```diff
@@ -0,0 +1,293 @@
+}
+
+var (
+	counts       = flag.String("counts", "1", "replica counts, must have at least one (csv)")
```
It looks like the assumption is that counts and times must be the same len(), but that is not called out.

uluyol replied:
Done.

thockin commented on contrib/diurnal/dc.go:
```diff
@@ -0,0 +1,293 @@
+	counts       = flag.String("counts", "1", "replica counts, must have at least one (csv)")
+	times        = flag.String("times", "0", "times at which to set replica counts relative to UTC (csv)")
+	userLabels   = flag.String("labels", "", "replication controller labels, should be of form key=value (csv)")
+	namespace    = flag.String("ns", "default", "namespace of replication controller")
```
you should assume it's the same namespace you are running in.

uluyol replied:
Done.

thockin commented on contrib/diurnal/dc.go:
```diff
@@ -0,0 +1,293 @@
+	times        = flag.String("times", "0", "times at which to set replica counts relative to UTC (csv)")
+	userLabels   = flag.String("labels", "", "replication controller labels, should be of form key=value (csv)")
+	namespace    = flag.String("ns", "default", "namespace of replication controller")
+	maxScaleRate = flag.Float64("rate", 5, "max number of replicas that may be added/removed per minute")
```
the flag should be max-rate or something with "max"

uluyol replied:
Done.

thockin commented on contrib/diurnal/dc.go:
```diff
@@ -0,0 +1,293 @@
+	maxScaleRate = flag.Float64("rate", 5, "max number of replicas that may be added/removed per minute")
+	startNow     = flag.Bool("now", false, "times are relative to now not 0:00 UTC (for demos)")
+	local        = flag.Bool("local", false, "set to true if running on local machine not within cluster")
+	proxyPort    = flag.Int("proxyport", 8001, "port that kubectl proxy is running on (local must be true)")
```
I'd just call this local-port

uluyol replied:
Done.

thockin commented on contrib/diurnal/dc.go:
```diff
@@ -0,0 +1,293 @@
+	return nil
+}
+
+func parseLabelsSelector(csvLabelSelector string) (labels.Selector, error) {
```
don't we have a reusable function for this?

thockin replied:
pkg/labels Parse() maybe?

uluyol replied:
Switched to using labels.Parse.

thockin commented on contrib/diurnal/dc.go:
```diff
@@ -0,0 +1,293 @@
+}
+
+func (s *Scaler) timeOffset() time.Duration {
+	return time.Since(s.start) % (24 * time.Hour)
```
Define a const dayPeriod = 24*time.Hour or similar

uluyol replied:
Done.

thockin commented on contrib/diurnal/dc.go:
```diff
@@ -0,0 +1,293 @@
+	return &Scaler{timeCounts: smoothed, selector: selector}, nil
+}
+
+// smoothChanges will add intermediate changes so that the rate of change will not
```
Is the intent just to stay below maxRate or to actually interpolate between times smoothly?

uluyol replied:
Stay below maxRate. The description has been updated.

uluyol replied:
Removing this in the next version. Will do simple jumping between replica counts.

uluyol replied:
Done.

thockin commented on contrib/diurnal/dc.go:
```diff
@@ -0,0 +1,293 @@
+
+var (
+	counts       = flag.String("counts", "1", "replica counts, must have at least one (csv)")
+	times        = flag.String("times", "0", "times at which to set replica counts relative to UTC (csv)")
```
what units?  Seconds?  arbitrary strings?

erictune replied:
ISO 8601?

uluyol replied:
Done.

thockin commented on contrib/diurnal/dc.go:
```diff
@@ -0,0 +1,293 @@
+	var tc []timeCount
+	prev := time.Duration(-1)
+	for i := range ts {
+		t, err := time.ParseDuration(ts[i])
```
Duration?  That's not what I expected.  You need a few good examples of time-lists

uluyol replied:
Switched to ISO 8601 times.

thockin commented on contrib/diurnal/dc.go:
```diff
@@ -0,0 +1,293 @@
+
+type Scaler struct {
+	timeCounts []timeCount
+	ctlr       *api.ReplicationController
```
never used

uluyol replied:
Removed.

thockin commented on contrib/diurnal/dc_test.go:
```diff
@@ -0,0 +1,114 @@
+		err    bool
+	}{
+		{
+			"1s,2m,3h,4h", "1,4,1,8", []timeCount{
```
I would expect these to be absolute time-of-day values, rather than durations - are the cumulative or absolute?

uluyol replied:
Absolute, should be more obvious now.

thockin commented on contrib/diurnal/dc_test.go:
```diff
@@ -0,0 +1,114 @@
+			t.Errorf("case %d: smoothed list is too short, smoothed: %v", i, smoothed)
+		}
+	}
+}
```
There's more you could test - findPos, setCount, timeOffset, curPos...

thockin commented on contrib/diurnal/diurnal_math.sh:
```diff
@@ -0,0 +1,49 @@
+
+diurnal_args=()
+
+for ((i=1; i < $#; i++)); do
```
Why not something simpler like

```
while [[ $# -gt 0 ]; do
    if [[ "$1" == "-sample" ]]; then
        shift
        arg=$1
    fi
    shift
done

thockin commented on contrib/diurnal/diurnal_math.sh:
```diff
@@ -0,0 +1,49 @@
+
+Usage of diurnal_math:
+  -sample s: sample every s seconds. Must be a positive integer
+  -countfn f: generate replica counts using f where f is a function of t.
```
maybe clarify this is a python function.

examples go a long way

thockin commented:
Overall the code is good.  It's a cool little demo.  A bit of polish will make it more professional - the arbitrary functions is sort of "cute" but the time/value logic is actually really useful.

erictune commented on contrib/diurnal/example-diurnal-controller.yaml:
```diff
@@ -0,0 +1,27 @@
+            - name: MAX_HEAP_SIZE
+              value: 512M
+            - name: HEAP_NEWSIZE
+              value: 100M
```
Since when does it take 100M to calculate a sine wave?

uluyol-goog committed:
Add a simple diurnal controller.

The diurnal controller changes the number of replicas of a replication controller based on a list of times and replica counts. It is meant to be run under a replication controller.

uluyol commented:
Removed the math function support. Moved to a busybox base image since no python or bash is needed now. Switched the example to using times and counts with the removal of the math support. The expected time format is now more sane (HHMMSS, HH:MM:SS, HHMM, HH:MM, or HH). Previously, the intermediate steps would be created earlier than desired (at the beginning of the space between two counts rather than at the end), this has now been fixed.

uluyol commented on contrib/diurnal/dc.go:
```diff
@@ -0,0 +1,337 @@
+
+var epoch = timeMustParse("150405", "000000")
+
+func parseTimeRelative(s string) (time.Duration, error) {
```
Will change to ISO 8601

uluyol replied:
Done.

thockin commented on contrib/diurnal/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM busybox
+MAINTAINER Muhammed Uluyol "uluyol@google.com"
+
+ADD dc /usr/bin/diurnal
```
I mostly see people stick custom binaries in /, FWIW - easier to find

uluyol replied:
Done.

thockin commented on contrib/diurnal/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+
```
dead line

uluyol replied:
Removed.

thockin commented on contrib/diurnal/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+ADD dc /usr/bin/diurnal
+
+RUN chown root:users /usr/bin/diurnal
+RUN chmod 755 /usr/bin/diurnal
```
you probably want an ENTRYPOINT directive

uluyol replied:
Done.

thockin commented on contrib/diurnal/Makefile:
```diff
@@ -0,0 +1,24 @@
+.PHONY: build push vet test clean
+
+TAG = 0.4.2
```
minor: it's simpler to keep tags to 1 or 2 digits unless you're seriously doing development on this.  v0.4, v0.5...v1.0

uluyol replied:
Done.

thockin commented on contrib/diurnal/README.md:
```diff
@@ -0,0 +1,39 @@
+This controller manipulates the number of replicas maintained by a replication controller throughout the day based on a provided list of times of day and replica counts. It should be run under a replication controller that is in the same namespace as the replication controller that it is manipulating.
+
+Instead of providing replica counts and times of day directly, you may use a script like the one below to generate them using mathematical functions.
+
```
Add an example of what a list of these looks like.

Add a link to the example YAML

uluyol replied:
Done.

thockin commented on contrib/diurnal/dc.go:
```diff
@@ -0,0 +1,282 @@
+	return t
+}
+
+var epoch = timeMustParse("150405", "000000")
```
comment on the magic numbers?

uluyol replied:
Done.

thockin commented on contrib/diurnal/example-diurnal-controller.yaml:
```diff
@@ -0,0 +1,27 @@
+kind: ReplicationController
+metadata:
+  labels:
+    name: diurnalcontroller
```
diurnal-controller ?

uluyol replied:
Done.

thockin commented on contrib/diurnal/example-diurnal-controller.yaml:
```diff
@@ -0,0 +1,27 @@
+        name: diurnalcontroller
+    spec:
+      containers:
+        - command: ["diurnal", "-labels", "name=redis-slave", "-times", "00:00Z,00:02Z,01:00Z,02:30Z", "-counts", "3,7,6,9"]
```
if you set ENTRYPOINT you can omit the first element here

uluyol replied:
Done.

thockin commented on contrib/diurnal/time.go:
```diff
@@ -0,0 +1,200 @@
+	setTrue
+)
+
+func parseTimeISO8601(s string) (time.Time, error) {
```
Add comments on WHY we have our own time parser

uluyol replied:
Done.

uluyol commented:
Pushed new version.

uluyol commented:
Fixed boilerplate.

thockin commented:
LGTM

uluyol commented:
Ran run-gendocs.sh, added analytics link.

mikedanese commented:
@k8s-bot ok to test

mikedanese commented:
@uluyol this needs a rebase to get shippable working.

uluyol commented:
Rebased.

mikedanese commented:
@k8s-bot ok to test
